File: docs/refman/refman.lua
local template = require("resty.template")
template.print = function(s) return s end

-- Blocks = {
--   <category> = {
--      type = "block" | "type",
--      info = { -- block
--          category = <category>,
--          name = <name>,
--          description = <description>,
--          params = {
--              {
--                  name = <name>,
--                  desc = <desc>,
--                  type = <type>,
--                  opt = <opt> or nil,
--              }, ...
--          }
--      }, ...
--   }, ...
-- }
local Blocks = {}

-- Modules = {
--   <module name> = {
--     {
--         type = "class" | "function" | "field",
--         info = { ... },
--     }, ...
--   }, ...
-- }
local Modules = {}

local function fix_whitespace(s)
    -- Trim beginning and ending whitespace
    s = string.gsub(s, "^%s*(.-)%s*$", "%1")
    -- Add leading newline to code fences (which gets mangled in the summary /
    -- description split by LDoc)
    s = string.gsub(s, "([^\n]) ```", "%1\n\n```")
    -- Trim first column of spaces after docstring comment
    s = string.gsub(s, "\n ", "\n")
    -- Add additional newline separate display mathjax
    s = string.gsub(s, "%$%$\n%$%$", "$$\n\n$$")

    return s
end

local function format_args_string(params)
    local required_args = {}
    local optional_args = {}

    for _, param in ipairs(params) do
        if param.opt == nil then
            required_args[#required_args + 1] = param.name
        elseif param.opt == true then
            optional_args[#optional_args + 1] = param.name
        else
            optional_args[#optional_args + 1] = param.name .. "=" .. tostring(param.opt)
        end
    end

    required_args = table.concat(required_args, ", ")
    optional_args = table.concat(optional_args, ", ")

    if #required_args > 0 and #optional_args > 0 then
        return required_args .. "[, " .. optional_args .. "]"
    elseif #required_args > 0 and #optional_args == 0 then
        return required_args
    elseif #optional_args > 0 and #required_args == 0 then
        return "[" .. optional_args .. "]"
    end

    return ""
end

local function extract_constructor(params, modifiers)
    local params_info = {}

    for i, param in ipairs(params or {}) do
        local name, desc = string.match(param, "([^%s]+)%s(.*)")
        local info = {
            name = name,
            desc = string.gsub(desc, "%s+%* ", "\n    * "),
            type = modifiers[i].type,
            opt = modifiers[i].opt,
        }
        params_info[#params_info + 1] = info
    end

    return params_info
end

local function extract_function(method, class_name)
    class_name = class_name or ""

    -- Extract arguments
    local method_args = {}
    for i, param in ipairs(method.params) do
        if method.modifiers.param[param] == nil then
            break
        end
        local info = {
            name = param,
            desc = fix_whitespace(method.params.map[param]),
            type = (method.modifiers.param[i] or {}).type,
            opt = (method.modifiers.param[i] or {}).opt,
        }
        method_args[#method_args + 1] = info
    end

    -- Extract returns
    local method_returns = {}
    for i, param in ipairs(method.ret or {}) do
        local info = {
            desc = param,
            type = method.modifiers['return'][i].type,
        }
        method_returns[#method_returns + 1] = info
    end

    -- Method info
    local info = {
        name = string.gsub(method.name, "[^%s]+([:.].*)", class_name .. "%1"),
        desc = fix_whitespace(method.summary .. method.description),
        args = method_args,
        args_string = format_args_string(method_args),
        returns = method_returns,
        raises = type(method.raise) == "string" and {method.raise} or method.raise,
        example = method.usage and fix_whitespace(method.usage[1]) or nil,
    }
    return info
end

local function extract_fields(table)
    -- Extract fields
    local fields_info = {}
    for _, param in ipairs(table.params) do
        local info = {
            name = param,
            desc = fix_whitespace(table.params.map[param]),
            type = table.modifiers.field[param].type,
        }
        fields_info[#fields_info + 1] = info
    end

    for subparam, _ in pairs(table.subparams) do
        for _, param in ipairs(table.subparams[subparam]) do
            local info = {
                name = param,
                desc = fix_whitespace(table.params.map[param]),
                type = table.modifiers.field[param].type,
            }
            fields_info[#fields_info + 1] = info
        end
    end

    return fields_info
end

local function lookup(module, type, name)
    for _, item in ipairs(Modules[module]) do
        if item.type == type and item.info.name == name then
            return item.info
        end
    end
end

local function format(t)
    for _, v in ipairs(t) do
        if v.tags.block then
            -- Extract constructor arguments
            local constructor_args = extract_constructor(v.tags.param, v.modifiers.param)

            -- Extract signature
            local signatures = {}
            for i, sig in ipairs(v.tags.signature) do
                signatures[i] = {inputs = {}, outputs = {}}
                local inputs, outputs = string.match(sig, "([^>]*)>([^>]*)")
                for name, type in string.gmatch(inputs, "([^:%s,]*):([^:%s,]*)") do
                    signatures[i].inputs[#signatures[i].inputs + 1] = {name = name, type = type}
                end
                for name, type in string.gmatch(outputs, "([^:%s,]*):([^:%s,]*)") do
                    signatures[i].outputs[#signatures[i].outputs + 1] = {name = name, type = type}
                end
            end

            -- Block info
            local info = {
                category = v.tags.category[1],
                name = v.tags.block[1],
                description = fix_whitespace(v.summary .. v.description),
                args = constructor_args,
                args_string = format_args_string(constructor_args),
                signatures = signatures,
                example = fix_whitespace(v.usage[1]),
            }

            -- Add it to our Blocks table
            Blocks[info.category] = Blocks[info.category] or {}
            Blocks[info.category][#Blocks[info.category] + 1] = {type = "block", info = info}

            -- If this block has an associated data type
            if #v.sections > 0 and v.sections[1].tags.datatype then
                v = v.sections[1]

                -- Extract constructor arguments
                local constructor_args = extract_constructor(v.tags.param, v.modifiers.param)

                -- Datatype info
                local type_info = {
                    name = v.tags.datatype[1],
                    description = fix_whitespace(v.summary .. v.description),
                    args = constructor_args,
                    args_string = format_args_string(constructor_args),
                    methods = {},
                }

                -- Add it to our Blocks table
                Blocks[info.category] = Blocks[info.category] or {}
                Blocks[info.category][#Blocks[info.category] + 1] = {type = "type", info = type_info}
            end
        elseif v.tags.datatype then
            -- Datatype methods
            local methods_info = {}
            for _, method in pairs(v.items) do
                methods_info[#methods_info + 1] = extract_function(method, v.tags.datatype[1])
            end

            -- Extract constructor arguments
            local constructor_args = extract_constructor(v.tags.param, v.modifiers.param)

            -- Datatype info
            local info = {
                name = v.tags.datatype[1],
                description = fix_whitespace(v.summary .. v.description),
                args = constructor_args,
                args_string = format_args_string(constructor_args),
                methods = methods_info,
                example = v.usage and fix_whitespace(v.usage[1]) or nil
            }

            -- Add it to our modules table
            Modules['radio.types'] = Modules['radio.types'] or {}
            Modules['radio.types'][#Modules['radio.types'] + 1] = {type = "class", info = info}
        else
            local module_name = string.gsub(v.name, ".core", "")

            -- Collect class types
            local classes_info = {}
            for _, item in ipairs(v.sections) do
                if item.type == "type" then
                    -- Extract constructor arguments
                    local constructor_args = extract_constructor(item.tags.param, item.modifiers.param)

                    -- Class info
                    local info = {
                        name = item.name,
                        description = fix_whitespace(item.summary .. item.description),
                        args = constructor_args,
                        args_string = format_args_string(constructor_args),
                        methods = {},
                        example = item.usage and fix_whitespace(item.usage[1]) or nil
                    }
                    classes_info[#classes_info + 1] = info
                end
            end

            -- Collect methods for classes
            for _, class_info in ipairs(classes_info) do
                for _, item in ipairs(v.items) do
                    if item.type == "function" and item.name:match("(.*):.*") == class_info.name then
                        class_info.methods[#class_info.methods + 1] = extract_function(item, class_info.name)
                    end
                end
            end

            -- Collect function types
            local functions_info = {}
            for _, item in ipairs(v.items) do
                if item.type == "function" and not item.name:find("[:.]") then
                    functions_info[#functions_info + 1] = extract_function(item)
                end
            end

            -- Collect field types
            local fields_info = {}
            for _, item in ipairs(v.items) do
                if item.type == "table" then
                    fields_info = extract_fields(item)
                end
            end

            -- Add it to our modules table
            Modules[module_name] = Modules[module_name] or {description = fix_whitespace(v.summary .. v.description)}
            for _, info in ipairs(classes_info) do
                Modules[module_name][#Modules[module_name] + 1] = {type = "class", info = info}
            end
            for _, info in ipairs(functions_info) do
                Modules[module_name][#Modules[module_name] + 1] = {type = "function", info = info}
            end
            for _, info in ipairs(fields_info) do
                Modules[module_name][#Modules[module_name] + 1] = {type = "field", info = info}
            end
        end
    end
    local s = template.render("refman.template.md", {Blocks = Blocks, Modules = Modules, lookup = lookup})
    print(s)
end

return {filter = format}

File: docs/refman/refman.py
import collections
import enum
import glob
import re
import sys

import mako.template
import mako.lookup


################################################################################
# File to Docstring Comments Parser
################################################################################


class FileParserState(enum.Enum):
    SEARCH = 1
    COMMENT = 2


def parse_file(filename):
    state = FileParserState.SEARCH

    comment = []
    comments = []

    with open(filename) as f:
        for line in f:
            line = line.strip()

            if state == FileParserState.SEARCH:
                if line == "---":
                    state = FileParserState.COMMENT
            elif state == FileParserState.COMMENT:
                match = re.match(r'\-\-[ ]?', line)
                if match:
                    comment.append(line[match.end():])
                else:
                    if comment:
                        comments.append(comment)
                        comment = []
                    state = FileParserState.SEARCH

    return comments


################################################################################
# Docstring Comments to Docstrings Parser
################################################################################


class DocstringParserState(enum.Enum):
    DESCRIPTION = 1
    TAG = 2


Docstring = collections.namedtuple('Docstring', ['description', 'tags'])


class DocstringTags:
    def __init__(self, tags):
        self._tags = tags

    def lookup(self, key):
        for (k, v) in self._tags:
            if k == key:
                return v
        raise KeyError("Unknown key \"{}\" in tags.".format(key))

    def lookup_optional(self, key):
        try:
            return self.lookup(key)
        except KeyError:
            return None

    def lookup_multiple(self, keys):
        return [(k, v) for (k, v) in self._tags if any([k.startswith(key) for key in keys])]

    def keys(self):
        return [re.match(r'([a-z]+)', k).group(1) for (k, v) in self._tags]

    def has(self, key):
        return key in self.keys()

    def __str__(self):
        return str(self._tags)

    def __repr__(self):
        return repr(self._tags)


def parse_comment(comment):
    state = DocstringParserState.DESCRIPTION
    tag = None
    tag_lines = []

    description = []
    tags = []

    for line in comment:
        is_tag = re.match(r'@[a-z]+.*$', line)

        if state == DocstringParserState.DESCRIPTION:
            # If this is a tag, stop collecting description
            if is_tag:
                state = DocstringParserState.TAG
            else:
                description.append(line)

        if state == DocstringParserState.TAG:
            # Add assembled tag to tags
            if tag and is_tag:
                tags.append((tag, tag_lines))
                tag, tag_lines = None, []

            # Process new tag
            if is_tag:
                tag, value = re.match(r'@([^\s]+)[\s]*(.*)', line).groups()
                if value:
                    tag_lines.append(value)
            else:
                tag_lines.append(line)

    # Add final assembled tag to tags
    if tag:
        tags.append((tag, tag_lines))

    # Trim trailing newlines in description
    if description and description[-1] == "":
        description.pop()

    # Trim trailing newlines in tag lines
    for (tag, lines) in tags:
        if lines and lines[-1] == "":
            lines.pop()

    # Join description lines
    description = "\n".join(description)

    # Join tag lines
    tags = [(tag, "\n".join(lines)) for (tag, lines) in tags]

    return Docstring(description, DocstringTags(tags))


def parse_comments(comments):
    docstrings = []

    for comment in comments:
        docstrings.append(parse_comment(comment))

    return docstrings


################################################################################
# Docstrings to Docs Decoder
################################################################################

# Assorted leaf elements

ParameterDoc = collections.namedtuple('ParameterDoc', ['name', 'type', 'default', 'description'])

ReturnDoc = collections.namedtuple('ReturnDoc', ['type', 'description'])

PortDoc = collections.namedtuple('PortDoc', ['name', 'type'])

FieldDoc = collections.namedtuple('FieldDoc', ['name', 'type', 'description'])

# Mid-level elements

SignatureDoc = collections.namedtuple('SignatureDoc', ['inputs', 'outputs'])

TableDoc = collections.namedtuple('TableDoc', ['name', 'description', 'children'])

FunctionDoc = collections.namedtuple('FunctionDoc', ['name', 'description', 'parameters', 'returns', 'raises', 'usage', 'static'])

PropertyDoc = collections.namedtuple('PropertyDoc', ['name', 'description', 'returns', 'raises', 'usage', 'static'])

# High-level elements

BlockDoc = collections.namedtuple('BlockDoc', ['name', 'description', 'category', 'parameters', 'signatures', 'usage', 'children'])

ModuleDoc = collections.namedtuple('ModuleDoc', ['name', 'description', 'children'])

ClassDoc = collections.namedtuple('ClassDoc', ['name', 'description', 'parameters', 'usage', 'children'])

DatatypeDoc = collections.namedtuple('DatatypeDoc', ['name', 'description', 'parameters', 'children'])


def decode_tag_parameters(tags):
    parameters = []

    # @tparam[opt=<default>] <type> <name> <description>
    # @param <name> <description>
    for (k, v) in tags:
        if k.startswith("tparam"):
            default = re.match(r'tparam\[opt=(.*)\]', k).group(1) if "opt=" in k else None
            type, name, description = re.match(r'([^\s]+) ([^\s]+) (.*)', v, re.MULTILINE | re.DOTALL).groups()
            parameters.append(ParameterDoc(name=name, type=type, default=default, description=description))
        elif k.startswith("param"):
            name, description = re.match(r'([^\s]+) (.*)', v, re.MULTILINE | re.DOTALL).groups()
            parameters.append(ParameterDoc(name=name, type=None, default=None, description=description))
        else:
            raise ValueError("Unknown parameter tag type \"{}\".".format(k))

    return parameters


def decode_tag_returns(tags):
    returns = []

    # @treturn <type> <description>
    # @return <description>
    for (k, v) in tags:
        if k == "treturn":
            type, description = re.match(r'([^\s]+) (.*)', v, re.MULTILINE | re.DOTALL).groups()
            returns.append(ReturnDoc(type=type, description=description))
        elif k == "return":
            description = v
            returns.append(ReturnDoc(type=None, description=description))
        else:
            raise ValueError("Unknown return tag type \"{}\".".format(k))

    return returns


def decode_tag_signatures(tags):
    signatures = []

    # @signature [<name>:<type>]... > [<name>:<type>]...
    for (k, v) in tags:
        if ">" not in v:
            raise ValueError("Invalid signature value \"{}\".".format(v))

        inputs, outputs = v.split(">")
        inputs, outputs = inputs.strip(), outputs.strip()

        input_ports, output_ports = [], []

        if inputs:
            for input in inputs.split(", "):
                if input == "...":
                    input_ports.append(PortDoc(name=input, type=None))
                else:
                    name, type = input.split(":")
                    input_ports.append(PortDoc(name=name, type=type))

        if outputs:
            for output in outputs.split(", "):
                if output == "...":
                    output_ports.append(PortDoc(name=output, type=None))
                else:
                    name, type = output.split(":")
                    output_ports.append(PortDoc(name=name, type=type))

        signatures.append(SignatureDoc(inputs=input_ports, outputs=output_ports))

    return signatures


def decode_tag_fields(tags):
    fields = []

    # @tfield <type> <name> <description>
    for (k, v) in tags:
        type, name, description = re.match(r'([^\s]+) ([^\s]+) (.*)', v, re.MULTILINE | re.DOTALL).groups()
        fields.append(FieldDoc(name=name, type=type, description=description))

    return fields


def decode_function_name(name):
    # <class>.<static function>
    # <class>:<method>
    # <module function>
    if "." in name:
        cls, static = name.split(".")[0], True
    elif ":" in name:
        cls, static = name.split(":")[0], False
    else:
        cls, static = None, True

    return (name, cls, static)


def decode_docstring_block(docstring):
    name = docstring.tags.lookup("block")
    description = docstring.description
    category = docstring.tags.lookup("category")
    parameters = decode_tag_parameters(docstring.tags.lookup_multiple(("tparam", "param")))
    signatures = decode_tag_signatures(docstring.tags.lookup_multiple(("signature",)))
    usage = docstring.tags.lookup("usage")

    return BlockDoc(name=name, description=description, category=category, parameters=parameters, signatures=signatures, usage=usage, children=[])


def decode_docstring_class(docstring):
    name = docstring.tags.lookup("class")
    description = docstring.description
    parameters = decode_tag_parameters(docstring.tags.lookup_multiple(("tparam", "param")))
    usage = docstring.tags.lookup_optional("usage")

    return ClassDoc(name=name, description=description, parameters=parameters, usage=usage, children=[])


def decode_docstring_datatype(docstring):
    name = docstring.tags.lookup("datatype")
    description = docstring.description
    parameters = decode_tag_parameters(docstring.tags.lookup_multiple(("tparam", "param")))

    return DatatypeDoc(name=name, description=description, parameters=parameters, children=[])


def decode_docstring_module(docstring):
    name = docstring.tags.lookup("module")
    description = docstring.description
    children = decode_tag_fields(docstring.tags.lookup_multiple(("tfield",)))

    return ModuleDoc(name=name, description=description, children=children)


def decode_docstring_function(docstring):
    name, cls, static = decode_function_name(docstring.tags.lookup("function"))
    description = docstring.description
    parameters = decode_tag_parameters(docstring.tags.lookup_multiple(("tparam", "param")))
    returns = decode_tag_returns(docstring.tags.lookup_multiple(("treturn", "return")))
    raises = [v for (_, v) in docstring.tags.lookup_multiple(("raise",))]
    usage = docstring.tags.lookup_optional("usage")
    static = static

    return FunctionDoc(name=name, description=description, parameters=parameters, returns=returns, raises=raises, usage=usage, static=static), cls


def decode_docstring_property(docstring):
    name, cls, static = decode_function_name(docstring.tags.lookup("property"))
    description = docstring.description
    returns = decode_tag_returns(docstring.tags.lookup_multiple(("treturn", "return")))
    raises = [v for (_, v) in docstring.tags.lookup_multiple(("raise",))]
    usage = docstring.tags.lookup_optional("usage")

    return PropertyDoc(name=name, description=description, returns=returns, raises=raises, usage=usage, static=static), cls


def decode_docstring_table(docstring):
    name = docstring.tags.lookup("table")
    description = docstring.description
    children = decode_tag_fields(docstring.tags.lookup_multiple(("tfield",)))

    return TableDoc(name=name, description=description, children=children)


def decode_docstrings(docstrings):
    docs = []
    namespace = None

    for docstring in docstrings:
        tag_keys = docstring.tags.keys()

        # Validate tags in this docstring
        for key in tag_keys:
            if key not in ['internal', 'block', 'module', 'class', 'datatype', 'function',
                           'property', 'table', 'tparam', 'param', 'treturn', 'return',
                           'tfield', 'raise', 'usage', 'category', 'signature']:
                raise ValueError("Unknown tag \"{}\" encountered in docstring: \"{}\"".format(key, docstring))

        if 'internal' in tag_keys:
            # Skip internal documentation
            pass
        elif 'block' in tag_keys:
            doc = decode_docstring_block(docstring)
            docs.append(doc)
            namespace = doc
        elif 'module' in tag_keys:
            doc = decode_docstring_module(docstring)
            docs.append(doc)
            namespace = doc
        elif 'class' in tag_keys:
            doc = decode_docstring_class(docstring)
            if namespace:
                namespace.children.append(doc)
            else:
                docs.append(doc)
                namespace = doc
        elif 'datatype' in tag_keys:
            doc = decode_docstring_datatype(docstring)
            if namespace:
                namespace.children.append(doc)
            else:
                docs.append(doc)
                namespace = doc
        elif 'function' in tag_keys:
            doc, cls = decode_docstring_function(docstring)

            parent = next(filter(lambda d: d.name == cls, docs + namespace.children), None)
            if parent:
                parent.children.append(doc)
            else:
                assert namespace, "Function docstring \"{}\" has no namespace.".format(doc.name)
                namespace.children.append(doc)
        elif 'property' in tag_keys:
            doc, cls = decode_docstring_property(docstring)

            parent = next(filter(lambda d: d.name == cls, docs + namespace.children), None)
            if parent:
                parent.children.append(doc)
            else:
                assert namespace, "Property docstring \"{}\" has no namespace.".format(doc.name)
                namespace.children.append(doc)
        elif 'table' in tag_keys:
            doc = decode_docstring_table(docstring)
            assert namespace, "Table docstring \"{}\" has no namespace.".format(doc.name)
            namespace.children.append(doc)
        else:
            raise ValueError("Unsupported docstring. Missing @block, @class, @module, @datatype, @function, @property, or @table tag. Docstring: \"{}\".".format(docstring))

    return docs


################################################################################
# Organization
################################################################################


def organize_docs(docs):
    blocks = {}
    modules = {}
    datatypes = {}

    for doc in docs:
        if isinstance(doc, BlockDoc):
            if doc.category not in blocks:
                blocks[doc.category] = []
            blocks[doc.category].append(doc)
        elif isinstance(doc, ModuleDoc):
            modules[doc.name] = doc
        elif isinstance(doc, DatatypeDoc):
            datatypes[doc.name] = doc
        else:
            raise ValueError("Unknown top-level doc type \"{}\".".format(type(doc)))

    for category in blocks:
        blocks[category] = sorted(blocks[category], key=lambda b: b.name)

    return blocks, modules, datatypes


################################################################################
# Debug dump
################################################################################


def dump(blocks, modules, datatypes):
    s = []

    s.append("Blocks")
    for category in blocks:
        s.append("\t" + category)
        for block in blocks[category]:
            s.append("\t\t{} ({})".format(block.name, type(block).__name__))
            if hasattr(block, "children"):
                for child in block.children:
                    s.append("\t\t\t{} ({})".format(child.name, type(child).__name__))

    s.append("Modules")
    for name in modules:
        s.append("\t{} ({})".format(name, type(modules[name]).__name__))
        for child in modules[name].children:
            s.append("\t\t{} ({})".format(child.name, type(child).__name__))
            if hasattr(child, "children"):
                for subchild in child.children:
                    s.append("\t\t\t{} ({})".format(subchild.name, type(subchild).__name__))

    s.append("Datatypes")
    for name in datatypes:
        s.append("\t{} ({})".format(name, type(datatypes[name]).__name__))
        for child in datatypes[name].children:
            s.append("\t\t{} ({})".format(child.name, type(child).__name__))

    return "\n".join(s)


################################################################################
# Top-level
################################################################################

if __name__ == "__main__":
    docs = []

    for filename in glob.iglob('../../radio/**/*.lua', recursive=True):
        # Skip thirdparty libraries
        if filename.startswith('../../radio/thirdparty/'):
            continue

        comments = parse_file(filename)
        docstrings = parse_comments(comments)
        docs += decode_docstrings(docstrings)

    # Organize docs into blocks, modules, datatypes categories
    blocks, modules, datatypes = organize_docs(docs)

    # Dump organized docs to stdout
    sys.stderr.write(dump(blocks, modules, datatypes) + "\n")

    # Templatize reference manual
    escape_markdown_headers = lambda s: re.sub(r"([#]+)", r'${"\1"}', s)
    refman_template = mako.template.Template(filename="refman.template.md",
                                             lookup=mako.lookup.TemplateLookup(directories=['./'],
                                                                               input_encoding='utf-8',
                                                                               preprocessor=escape_markdown_headers),
                                             preprocessor=escape_markdown_headers)

    sys.stdout.write(refman_template.render(blocks=blocks, modules=modules, datatypes=datatypes,
                                            ModuleDoc=ModuleDoc, BlockDoc=BlockDoc, DatatypeDoc=DatatypeDoc,
                                            ClassDoc=ClassDoc, FunctionDoc=FunctionDoc, PropertyDoc=PropertyDoc,
                                            FieldDoc=FieldDoc))
